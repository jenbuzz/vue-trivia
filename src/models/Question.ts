type Difficulty = 'easy' | 'medium' | 'hard';

export interface Question {
    difficulty: Difficulty;
    question: string;
    correct: string;
    incorrect: string[];
}
