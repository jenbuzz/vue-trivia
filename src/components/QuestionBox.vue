<template>
    <div>
        <p v-html="currentQuestion.question"></p>
        <b-list-group :class="[disableList && 'disabled']">
            <b-list-group-item
                v-for="(answer, index) in answers"
                :key="index"
                @click="setSelectedAnswer(index)"
                :class="getClassName(index)"
            >
                <span v-html="answer"></span>
            </b-list-group-item>
        </b-list-group>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Watch, Vue } from 'vue-property-decorator';
import { Question } from '../models/Question';

@Component
export default class QuestionBox extends Vue {
    @Prop() private readonly currentQuestion!: Question;
    @Prop() private nextQuestion!: Function;
    @Prop() private incrementCorrectAnswers!: Function;

    private answers: string[] | null = null;
    private selectedAnswer: number | null = null;
    private correctAnswer: number | null = null;
    private disableList = false;

    @Watch('currentQuestion', { immediate: true })
    onNewQuestion() {
        this.selectedAnswer = null;
        this.disableList = false;

        const answers = [...this.currentQuestion.incorrect, this.currentQuestion.correct];

        for (let i = answers.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * i);
            const tmp = answers[i];
            answers[i] = answers[j];
            answers[j] = tmp;
        }

        this.answers = answers;
        this.correctAnswer = answers.indexOf(this.currentQuestion.correct);
    }

    setSelectedAnswer(index: number) {
        this.selectedAnswer = index;
        this.disableList = true;

        if (this.selectedAnswer === this.correctAnswer) {
            this.incrementCorrectAnswers();
        }

        setTimeout(() => {
            this.nextQuestion();
        }, 500);
    }

    getClassName(index: number): string {
        if (index === this.selectedAnswer && index === this.correctAnswer) {
            return 'positive';
        }

        if (index === this.selectedAnswer && index !== this.correctAnswer) {
            return 'negative';
        }

        return '';
    }
}
</script>

<style scoped lang="scss">
p {
    font-weight: 600;
}

.list-group {
    border: 1px solid #222222;

    .list-group-item {
        border: 0;
        cursor: pointer;

        &:hover {
            background: #e2e2e2;
        }

        &.positive {
            background: #68d388;
            color: #ffffff;
        }

        &.negative {
            background: #ff0028;
            color: #ffffff;
        }
    }

    &.disabled {
        .list-group-item:not(.positive):not(.negative) {
            &:hover {
                background: none;
            }
        }
    }
}
</style>
