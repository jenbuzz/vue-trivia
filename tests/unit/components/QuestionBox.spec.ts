import { createLocalVue, shallowMount } from '@vue/test-utils';
import BootstrapVue from 'bootstrap-vue';
import QuestionBox from '@/components/QuestionBox.vue';
import { Question } from '@/models/Question';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

describe('Component: QuestionBox', () => {
    it('renders', () => {
        const question: Question = {
            difficulty: 'easy',
            question: 'Who painted the Mona Lisa?',
            correct: 'Leonardo da Vinci',
            incorrect: ['Pablo Picasso', 'Claude Monet', 'Vincent van Gogh'],
        };

        const wrapper = shallowMount(QuestionBox, {
            localVue,
            propsData: {
                currentQuestion: question,
                nextQuestion: jest.fn(),
                incrementCorrectAnswers: jest.fn(),
            },
            data: function() {
                return {
                    answers: [
                        'Leonardo da Vinci',
                        'Vincent van Gogh',
                        'Pablo Picasso',
                        'Claude Monet',
                    ],
                };
            },
            watch: {
                currentQuestion: jest.fn(),
            },
        });

        wrapper.vm.$nextTick(() => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
