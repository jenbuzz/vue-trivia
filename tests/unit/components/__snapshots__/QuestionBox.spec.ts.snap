// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: QuestionBox renders 1`] = `
<div>
  <p>Who painted the Mona Lisa?</p>
  <b-list-group-stub tag="div">
    <b-list-group-item-stub tag="div" target="_self" class=""><span>Leonardo da Vinci</span></b-list-group-item-stub>
    <b-list-group-item-stub tag="div" target="_self" class=""><span>Vincent van Gogh</span></b-list-group-item-stub>
    <b-list-group-item-stub tag="div" target="_self" class=""><span>Pablo Picasso</span></b-list-group-item-stub>
    <b-list-group-item-stub tag="div" target="_self" class=""><span>Claude Monet</span></b-list-group-item-stub>
  </b-list-group-stub>
</div>
`;
